src/net_processing.cpp

```diff
@@ -3988,11 +3988,56 @@ void PeerLogicValidation::ConsiderEviction(CNode& pto, int64_t time_in_seconds)
 
 void PeerLogicValidation::EvictExtraOutboundPeers(int64_t time_in_seconds)
 {
-    // Check whether we have too many outbound peers
-    int extra_peers = m_connman.GetExtraOutboundCount();
+    // If we have any extra block-relay peers, disconnect the youngest unless
+    // it's given us a block -- in which case, disconnect the second youngest
+    // instead (so we rotate our youngest block-relay peer).
+    int extra_block_relay_peers = m_connman.GetExtraBlockRelayCount();
+    if (extra_block_relay_peers > 0) {
+        // Pick the block-relay peer that we just connected, unless it gave us
+        // a block -- in which case disconnect the next-youngest block-relay
+        // peer.
+        std::pair<NodeId, int64_t> youngest_peer = {-1, std::numeric_limits<int64_t>::max()};
+        std::pair<NodeId, int64_t> second_youngest_peer = {-1, std::numeric_limits<int64_t>::max()};
+
+        m_connman.ForEachNode([&](CNode* pnode) {
+            LockAssertion lock(::cs_main);
+            if (!pnode->IsBlockOnlyConn() || pnode->fDisconnect) return;
+            if (pnode->GetId() > youngest_peer.first) {
```

Comment:
This is making an assumption on `GetNewNodeId()` being a monotonic counter function of connections order. It may silently break id we modify ids selection to something else (like random nonces). Can we use `nTimeConnected` instead ?

Reply:
I'd also prefer to not having this assumption, and it seems to be easy to avoid.

Reply:
If `m_connman.ForEachNode` is not ordered, this algorithm may mistakenly select `second_youngest`.

Reply:
Sure I can change this, but keep in mind this logic is used a little further down already, in the existing outbound full-relay peer eviction algorithm.

Reply:
As alternative have you considered caching them ? We know both when we open such connections and when we drop them. It would avoid the risk of logic bug and iterating every other connections types not concerned by such eviction.

I think it's more a future work direction.

Reply:
I assume you mean caching at the time we open the connection?  I think that is problematic, because in order to keep that extra state up to date in the event that peers get disconnected and we make new connections after that, you have to a lot of additional error checking and introduced added complexity.  Doing all the checks in one place, at the point in time that we care about getting to a new/correct state when we're over our maximum, seems simpler to me to reason about.